### ISSUE TITLE
Support concurrent --show while cracking with a -ztex format

### ISSUE THREAD

**Issue by solardiz** [magnumripper/JohnTheRipper#3709]
The "device busy" conflict identified in #3379 also prevents "--show" from working while cracking. I have to explicitly specify e.g. "--show --format=descrypt" to override autodetection as "descrypt-ztex". It's unlikely our users would guess to do that. But this gave me an idea: maybe we can have john.c skip registration of the -ztex formats when the invocation is for "--show" (check that flag)?

Alternatively, maybe we can have -ztex formats themselves not access the actual devices when used in "--show".

**Comment by solardiz** [comment]
Tested after #3713, works. Thanks!

**Comment by AlekseyCherepanov** [comment]
- It has a minor side-effect: if user had -ztex format specified explicitly and changed command to add `--show`, then the command with `--format=...-ztex` and `--show` would fail with `Unknown ciphertext format name requested`.
- The implemented change does not affect `--show=left`, I see `ZTEX [...] bus:4 dev:4 Frequency:180 180 180 180` before hashes.

**Comment by solardiz** [comment]
Thanks, Aleksey. I think we can live with this minor side-effect, but it's totally weird and unnecessary that `--show=left` isn't covered by the fix for this issue. So I'll re-open the issue and expect @Apingis to at least look into fixing that.

As to the minor side-effect, maybe it too can be avoided by not skipping the ZTEX formats registration when any `--format` option is given (check that option's flag). @Apingis please look into that while you're at it, too. Thanks!

**Comment by solardiz** [comment]
I think ZTEX formats may postpone device initialization on `init` and `reset`, but instead set a global flag (a static variable) requesting that the device initialization function be called whenever the device would actually be needed (perhaps on the next call to `crypt_all`).

As to Johnny, I think this is no big deal for now. There are few users of *-ztex formats and few users of Johnny. I think there are no users of both at once. But of course it's good to keep these compatible, especially given that we intend to merge Johnny into the same tree with jumbo.

**Comment by magnumripper** [comment]
I think we don't currently have any ztex or opencl formats that hasn't got a CPU equivalent, so at least you'd know a possible format and hopefully know there's also ztex support for it.

Ideally eg. md5crypt should be one single format, capable of launching CPU threads, ztex and/or opencl devices concurrently. So it would be `--format=md5crypt` and other (global) options would select what resources to use. But we'll probably never reach that.

**Comment by magnumripper** [comment]
For future enhancements I think we'll need to specify our format interface harder. I know we've got a description fairly recently of where/how ztex formats are initializing stuff but I can't remember where right now. Anyhow, maybe we should specify that things like initializations and actual "locking in" on hardware ought to be done no earlier than in `reset()` like in most or all OpenCL formats. That should get rid of these problems and some other. If `--show=left` does actually call reset, that in itself is a bug (actually as we know, the whole `--show=left` thing is a fugly hack so maybe it does do that - we need to rewrite it).

**Comment by solardiz** [comment]
@Apingis This issue still has some probably trivial work for you as I described in https://github.com/magnumripper/JohnTheRipper/issues/3709#issuecomment-478151412

**Comment by Apingis** [comment]
It could be better if it was more clear from specs in ```formats.h``` that ```init()``` is not necessarily followed by any cracking.
I'm considering moving hardware initialization to ```reset()```. There's an important difference that ```init()``` is called only once for a given format and ```reset()``` can be called several times. It's going to check hardware state (e.g. which bitstream loaded) on ```reset()```, maybe store more of its state internally.

Other thing is printing a list of devices. I'm going to separate that from initialization, considering to print the list only once per run.
